# Clase 3: Listas y Arreglos

## Listas

In [1]:
[10, 15, 20, 25]

[10, 15, 20, 25]

In [2]:
['Pedro', 'Pablo', 'Andrés']

['Pedro', 'Pablo', 'Andrés']

In [4]:
ejemplo = ['Pedro', 10, ['hola', 1.12]]

In [5]:
print(ejemplo)

['Pedro', 10, ['hola', 1.12]]


In [13]:
len(ejemplo)

3

### Obtener y modificar elementos

In [6]:
ejemplo[0]

'Pedro'

In [7]:
ejemplo[-1]

['hola', 1.12]

Cuidado, lo siguiente muestra el <b>tercer</b> elemento, no el segundo

In [8]:
ejemplo[2]

['hola', 1.12]

In [9]:
ejemplo[1] = 'adiós'

In [10]:
ejemplo

['Pedro', 'adiós', ['hola', 1.12]]

In [26]:
ejemplo[3]

IndexError: list index out of range

### Atravesar la lista

In [11]:
for element in ejemplo:
    print(element)

Pedro
adiós
['hola', 1.12]


In [12]:
numeros = [12,13,15,22,23]

In [14]:
for i in range(len(numeros)):
    numeros[i] = numeros[i] + 1

In [15]:
numeros

[13, 14, 16, 23, 24]

### Tajadas

In [22]:
numeros3[1:2]

[14]

In [23]:
numeros3[1:4]

[14, 16, 23]

In [25]:
numeros3[1:]

[14, 16, 23, 24, 1, 2, 3, 4, 5]

In [27]:
numeros3[:-2]

[13, 14, 16, 23, 24, 1, 2, 3]

In [28]:
numeros3[:]

[13, 14, 16, 23, 24, 1, 2, 3, 4, 5]

### Operaciones

In [35]:
numeros2 = [1,2,3,4,5]

In [20]:
numeros3 = numeros + numeros2

In [21]:
numeros3

[13, 14, 16, 23, 24, 1, 2, 3, 4, 5]

In [19]:
[0,1]*4

[0, 1, 0, 1, 0, 1, 0, 1]

In [36]:
numeros2.append(1)

In [37]:
numeros2

[1, 2, 3, 4, 5, 1]

In [38]:
numeros2.extend([1,2])

In [39]:
numeros2

[1, 2, 3, 4, 5, 1, 1, 2]

In [41]:
sum(numeros2)

19

In [42]:
del numeros2[1]

In [43]:
numeros2

[1, 3, 4, 5, 1, 1, 2]

## Tuplas

Las tuplas se parecen a las listas, pero no se pueden modificar

In [54]:
mi_tupla = (1,2,3)

In [55]:
mi_tupla[1]

2

In [56]:
mi_tupla[1] = 4

TypeError: 'tuple' object does not support item assignment

## Arreglos

Los arreglos son mucho más rápidos para cálculos numéricos ya que son un bloque continuo en memoria. Son similares a los arreglos en lenguajes como C.

In [44]:
import numpy as np

In [45]:
mi_arreglo = np.array([1,2,3])

In [72]:
mi_arreglo[1] = 5

In [73]:
mi_arreglo

array([1, 5, 3])

Los arreglos sólo pueden contener datos de un tipo de datos. Cuando se mezclan tipos, Python hará lo mejor que pueda para convertirlos a un úunico tipo.

In [46]:
np.array([1,'hola'])

array(['1', 'hola'], dtype='<U21')

Podemos hacer arreglos bidimensionales (matrices), tridimensionales, etc

In [51]:
np.array([[1,1],[1,1]])

array([[1, 1],
       [1, 1]])

Los arreglos no pueden contener elementos de dimensiones diferentes

In [50]:
np.array([1,'hola',[2,'adios']])

ValueError: setting an array element with a sequence

### Llenar arreglos automáticamente

In [57]:
np.ones([3,4])

array([[1., 1., 1., 1.],
       [1., 1., 1., 1.],
       [1., 1., 1., 1.]])

In [53]:
np.zeros((2,3,4))

array([[[0., 0., 0., 0.],
        [0., 0., 0., 0.],
        [0., 0., 0., 0.]],

       [[0., 0., 0., 0.],
        [0., 0., 0., 0.],
        [0., 0., 0., 0.]]])

In [67]:
np.full((3,2),3, dtype=np.float128)

array([[3., 3.],
       [3., 3.],
       [3., 3.]], dtype=float128)

In [68]:
np.arange(10,25,5)

array([10, 15, 20])

In [70]:
mi_arreglo2 = np.linspace(0,2,9)

In [71]:
mi_arreglo2

array([0.  , 0.25, 0.5 , 0.75, 1.  , 1.25, 1.5 , 1.75, 2.  ])

### Inspeccionar arreglos

In [74]:
mi_arreglo2.ndim

1

In [79]:
np.full((3,2),3, dtype=np.float128).size

6

In [77]:
mi_arreglo2.data

In [78]:
mi_arreglo2.shape

(9,)

In [80]:
mi_arreglo2.dtype

dtype('float64')

In [81]:
mi_arreglo2.nbytes

72

In [85]:
mi_arreglo2.astype(np.float16)

array([0.  , 0.25, 0.5 , 0.75, 1.  , 1.25, 1.5 , 1.75, 2.  ],
      dtype=float16)

In [86]:
mi_arreglo2.astype(np.float16).nbytes

18

### Aritmética de arreglos

In [87]:
mi_arreglo

array([1, 5, 3])

In [89]:
mi_arreglo3 =  mi_arreglo + 1

In [90]:
mi_arreglo3

array([2, 6, 4])

In [91]:
mi_arreglo + mi_arreglo3

array([ 3, 11,  7])

In [92]:
mi_arreglo3/mi_arreglo

array([2.        , 1.2       , 1.33333333])

In [119]:
mi_arreglo3 == mi_arreglo

array([False, False, False])

Se pueden usar los operadores +, -, *, /, %, **, ==, !=, >, <, >=, <=

In [93]:
mi_arreglo3.max()

6

In [94]:
mi_arreglo3.min()

2

### Tajadas de arreglos

In [101]:
mi_arreglo_2d = np.array([[1,2,3],[4,5,6],[7,8,9]])

In [102]:
mi_arreglo_2d

array([[1, 2, 3],
       [4, 5, 6],
       [7, 8, 9]])

In [103]:
mi_arreglo_2d[1:2]

array([[4, 5, 6]])

In [108]:
mi_arreglo_2d[1:]

array([[4, 5, 6],
       [7, 8, 9]])

In [111]:
mi_arreglo_2d[1,1]

5

¿Cuál es la diferencia? El siguiente código primero toma todas las filas desde la segunda (índice 1), luego toma la segunda de esas listas.

In [132]:
mi_arreglo_2d[1:][1]

array([7, 8, 9])

Por otra parte, el siguiente código toma el primer elemento de cada una de esas listas.

In [130]:
mi_arreglo_2d[1:,1]

array([5, 8])

In [114]:
mi_arreglo_2d[1:,:-1]

array([[4, 5],
       [7, 8]])

In [115]:
mi_arreglo

array([1, 5, 3])

In [118]:
mi_arreglo[mi_arreglo<=3]

array([1, 3])

In [120]:
menor_que_3 = mi_arreglo < 3

In [121]:
menor_que_3

array([ True, False, False])

In [122]:
mi_arreglo[menor_que_3]

array([1])

## Tareas

### Tarea 2.1

Escriba un código que tome un arreglo lleno de ceros, producido con np.zeros((3,2)) y usando ciclos for lo llena para obtener 

### Tarea 2.2

Escriba un código que tome un array producido con np.zeros(100) y produzca el mismo resultado que la función np.linspace(0,0.1,10), sin usar esa función.

In [145]:
np.linspace(0.1,10,100)

array([ 0.1,  0.2,  0.3,  0.4,  0.5,  0.6,  0.7,  0.8,  0.9,  1. ,  1.1,
        1.2,  1.3,  1.4,  1.5,  1.6,  1.7,  1.8,  1.9,  2. ,  2.1,  2.2,
        2.3,  2.4,  2.5,  2.6,  2.7,  2.8,  2.9,  3. ,  3.1,  3.2,  3.3,
        3.4,  3.5,  3.6,  3.7,  3.8,  3.9,  4. ,  4.1,  4.2,  4.3,  4.4,
        4.5,  4.6,  4.7,  4.8,  4.9,  5. ,  5.1,  5.2,  5.3,  5.4,  5.5,
        5.6,  5.7,  5.8,  5.9,  6. ,  6.1,  6.2,  6.3,  6.4,  6.5,  6.6,
        6.7,  6.8,  6.9,  7. ,  7.1,  7.2,  7.3,  7.4,  7.5,  7.6,  7.7,
        7.8,  7.9,  8. ,  8.1,  8.2,  8.3,  8.4,  8.5,  8.6,  8.7,  8.8,
        8.9,  9. ,  9.1,  9.2,  9.3,  9.4,  9.5,  9.6,  9.7,  9.8,  9.9,
       10. ])

### Tarea 2.3

A partir del siguiente array, representando una matriz, escriba un código produzca un nuevo sólo con los elementos de la segunda y tercera columna de cada fila, sin usar ciclos for ni while.

In [151]:
mi_array = np.array([[1,1,1,1,3],[1,1,1,1,5],[2,3,4,5,6],[2,2,2,2,8],[1,2,3,4,5]])

### Tarea 2.4

Escriba un código, sin usar ciclos for ni while, que calcule los cuadrados de los números enteros enttre 1 y 100.

### Tarea 2.5

Sin usar ciclos for ni while, escriba un código que calcule la siguiente sumatoria

$$
\sum_{n = 1}^{100} n*(n+1)
$$